src/libraries/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs

```diff
@@ -9,88 +10,190 @@ internal struct FileStatus
     {
         private const int NanosecondsPerTick = 100;
 
+        // The last cached lstat information about the file
+        private Interop.Sys.FileStatus _fileCache;
+
+        // -1: if the file cache isn't initialized - Refresh should always change this value
+        //  0: if the file cache was initialized with no errors
+        // Positive number: the error code returned by the lstat call
+        private int _initializedFileCache;
+
         // The last cached stat information about the file
-        private Interop.Sys.FileStatus _fileStatus;
+        // Refresh only collects this if lstat determines the path is a symbolic link
+        private Interop.Sys.FileStatus _symlinkCache;
```

Comment:
As we discussed offline, could it be better to not cache the whole Interop.Sys.FileStatus for stat(symlink) and instead just cache the result of `FileStatus.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFDIR`? Or perhaps it doesn't make a difference.

Reply:
I addressed the other suggestions you provided, ran benchmarks again, and noticed the ratio got more stable, all while keeping the symlink cache code.